File: ValidateBST_LC98.java

class Solution {
    List<Integer> res = new ArrayList<>();
    public boolean isValidBST(TreeNode root) {
        //if the inorder traversal of the tree is sorted it is a binary sub tree - 
        //do the inorder traversal and add the node in the result list then traverse the list and check if its sorted ascending order 
        //TC:O(n)
        //SC:O(n)
        helper(root);
        for(int i=1;i<res.size();i++){
            //System.out.print(res.get(i));
            if(res.get(i)<=res.get(i-1)){
                return false;
            }
        }
        return true;
    }
    
    private  List<Integer> helper(TreeNode root){
        
        if(root==null){
            return res;
        }
        
        helper(root.left);
        res.add(root.val);
        helper(root.right);
        
        return res;
    
    }
    
}
 */

//Approach 2:
 //if the inorder traversal of the tree is sorted it is a binary sub tree - 
        //while doing inorder traversal check if prev node is lesser than current node
        
        //TC:O(n)
        //SC:O(1)

// class Solution {
//    TreeNode prev;
//     boolean flag;
//     public boolean isValidBST(TreeNode root) {
//         //set default value of prev and flag
//         prev = null;
//         flag = true;
//         helper(root);
//         return flag;
//     }
    
//     private void helper(TreeNode root){
        
//         if(root==null){
//             return;
//         }
        
//         helper(root.left);
//       //check if the prev node was less than curr node , if not set flag to false
//         if(prev!=null && prev.val>=root.val){
//             flag = false;
//         }
//         prev=root;
//         //if flag is true then check the right subtree -- can apply same for left rec call
//         if(flag){
//         helper(root.right);
//         }
    
//     }
    
// }

//Approach 3:
 //if the iterative inorder traversal of the tree is sorted it is a binary sub tree - 
        //while doing inorder traversal check if prev node is lesser than current node
        
        //TC:O(n)
        //SC:O(1)

class Solution {
    public boolean isValidBST(TreeNode root) {
       
        TreeNode prev=null;
        Stack<TreeNode> s = new Stack<>();
        while(root!=null ||!s.isEmpty()){
           while(root!=null){
               s.push(root);
               root = root.left;
           } 
            root = s.pop();
            if(prev!= null && prev.val>=root.val)
            {
                return false;
            } 
            prev=root;
            root = root.right;
                
            
        }
       return true; 
    }    
}

File: validateBST-LC 98.py
# Time Complexity = O(n)
# Space Complexity = O(h), where h is the height of the tree, because at one go, h elements are being pushed in the stack


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    # Recursive Code to Check if it is a Valid BST
    # Logic - Check if the in order traversal is in ascending order. Only the previous and current node is required to compare
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        self.prev = None            # Initialising the previous node
        self.isValid = True
        
        if root is None:
            return True
        
        self.inorder(root, self.prev)    
    
        return self.isValid
    

    
    def inorder(self, root, prev):
       
        if root == None:
            return
        self.inorder(root.left, self.prev)      # Left Subtree Traversal
        # print("left", root.val)
        
        # Boundary condition to avoid errors in a single node BST (self.prev != None)
        if self.prev != None and self.prev.val >= root.val:
            self.isValid = False
            return
        
        self.prev = root                        # Moving the pointer ahead
        self.inorder(root.right, self.prev)     # Right Subtree Traversal
        # print("right", root.val)
        


# Iterative Code to print inorder traversal
'''
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
    
        stack = []
        
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
                
            root = stack.pop()
            print(root.val)
            root = root.right
            
'''        
        
    